lib/mongo/server/connection_pool.rb

```diff
@@ -73,19 +119,79 @@ def checkin(connection)
       #
       # @since 2.0.0
       def checkout
-        queue.dequeue
+        raise_if_closed!
+
+        publish_cmap_event(
+          Monitoring::Event::ConnectionCheckoutStarted.new(address)
+        )
+
+        deadline = Time.now + wait_timeout
+        @wait_queue.enter_wait_queue(wait_timeout, deadline) do
+          connections.pop(deadline).tap do |c|
+            publish_cmap_event(
+              Monitoring::Event::ConnectionCheckedOut.new(address, c.id),
+            )
+          end
+        end
+      rescue Error::WaitQueueTimeout
+        publish_cmap_event(
+            Monitoring::Event::ConnectionCheckoutFailed.new(
+              Monitoring::Event::ConnectionCheckoutFailed::TIMEOUT,
+              address,
+            ),
+        )
+        raise
       end
 
       # Disconnect the connection pool.
       #
-      # @example Disconnect the connection pool.
+      # @example Disconnect connection pool.
       #   pool.disconnect!
       #
       # @return [ true ] true.
       #
       # @since 2.1.0
       def disconnect!
-        queue.disconnect!
+        unless closed?
+          connections.disconnect!
+
+          publish_cmap_event(
+            Monitoring::Event::PoolClosed.new(address)
+          )
+        end
+
+        true
+      end
+
+      # Updates the generation number. The connections will be disconnected and removed lazily
+      # when the queue attempts to dequeue them.
+      #
+      # @return [ true ] true.
+      #
+      # @since 2.7.0
+      def clear!
+        unless closed?
```

Comment:
I think `clear` should raise if the pool is closed.

Reply:
Okay, I'll change it

Reply:
Done.